[ISSUE#1881][MAS2.1.2][Keyboard Navigation- Welcome Tab screen] Focus gets trapped on first and last element of the page when user navigates through the page

ceciliaavila opened the pull request:
Solves # 1881

### Description
Fixes how the focus gets trapped between the last and the first element of the pane.
 
### Changes made
We added a new condition to the **_EventHandlers_** class which will manage the focus when the emulator is running in _macOS_. Also, we included two auxiliary methods to find the first and last elements of the active pane and be able to move the focus between them.

Additionally, we added two unit tests to cover the new _keydown_ event.

### Testing
In the following image, you can see how the focus goes from the last element of the pane to the first when navigating with Tab key and vice-versa when using Shift+Tab navigation: 

![FocusTrap](https://user-images.githubusercontent.com/44245136/69159340-876b9b00-0ac6-11ea-857b-e76f10ead6f9.gif)

Aliandi commented on packages/app/client/src/utils/eventHandlers.ts:
```diff
@@ -35,13 +35,36 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
+      var child = this.getLastChildWithChildren(node);
+      if (child) {
+        list.push(child);
+        this.getLastDecendants(child, list);
```
Couldn't this all be solved with a map?

Aliandi replied:
`list = node.children.map(function(node) { this.getLastChildwithChildren(node) })`

Aliandi replied:
should probably filter after this

ceciliaavila replied:
We tried it but didn't work. Thank you!!

Aliandi commented on packages/app/client/src/utils/eventHandlers.ts:
```diff
@@ -35,13 +35,36 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
+  }
+
+  public static getLastDecendants(node, list = []) {
+    list = list || [];
```
Also, why this? list should be either something or the empty list already, this looks redundant

ceciliaavila replied:
Feedback applied. Thanks!

Bill7zz commented:
I suggest changing `we added a new keydown event for the Tab key` with `we added a new condition to the eventHandlers class which will manage the focus when the emulator is running in Mac OS`

ceciliaavila commented:
> I suggest changing `we added a new keydown event for the Tab key` with `we added a new condition to the eventHandlers class which will manage the focus when the emulator is running in Mac OS`

Done. Thanks!

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+
+  it('should move focus to first element when Tab is pressed', async () => {
+    const event = new KeyboardEvent('keydown', { key: 'Tab' });
+    Object.defineProperty(process, 'platform', { value: 'darwin' });
```
There'd be a space between variables declarations and the rest of the code.

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+    Object.defineProperty(process, 'platform', { value: 'darwin' });
+
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
```
Change both var to const

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
+    var mockLastElement = document.getElementById('btn3');
+    mockLastElement.focus();
```
add space between line 276 and 277

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+
+  it('should move focus to last element when Shift+Tab is pressed', async () => {
+    const event = new KeyboardEvent('keydown', { shiftKey: true, key: 'Tab' });
+    Object.defineProperty(process, 'platform', { value: 'darwin' });
```
Add space here

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
+    var mockLastElement = document.getElementById('btn3');
```
both var should be const

denscollo commented on packages/app/client/src/utils/eventHandlers.spec.ts:
```diff
@@ -253,4 +266,30 @@ describe('#globalHandlers', () => {
+    document.body.innerHTML = mockDOM;
+    var mockFirstElement = document.getElementById('navBtn');
+    var mockLastElement = document.getElementById('btn3');
+    mockFirstElement.focus();
```
add space between line 289 and 290

denscollo commented on packages/app/client/src/utils/eventHandlers.ts:
```diff
@@ -35,18 +35,41 @@ import { Notification, NotificationType, SharedConstants } from '@bfemulator/app
+
+  public static getLastDecendants(node, list = []) {
+    if (node.children.length > 0) {
+      var child = this.getLastChildWithChildren(node);
```
change var to const

denscollo commented on packages/app/client/src/utils/eventHandlers.ts:
```diff
@@ -121,6 +144,27 @@ class EventHandlers {
     }
+
+    if (isMac()) {
+      var KEYCODE_TAB = 9;
```
All these variables should be const
